source/CIS/cis-4.rst

```diff
@@ -0,0 +1,791 @@
+.. _CIS-4:
+
+===================================
+CIS-4: Credential Registry Standard
+===================================
+
+.. list-table::
+   :stub-columns: 1
+
+   * - Created
+     - May 1, 2023
+   * - Supported versions
+     - | Smart contract version 1 or newer
+       | (Protocol version 4 or newer)
+   * - Standard identifier
+     - ``CIS-4``
+
+
+Abstract
+========
+
+A standard interface for verifiable credential (VC) registries.
+A registry keeps track of public VC data and manages the VC lifecycle.
+The interface provides functionality for the following roles of users:
+
+- *credential issuers* can register and revoke credentials, manage revocation keys;
+- *credential holders* can revoke holder-revocable credentials by signing a revocation message;
+- *verifiers* can query credential status and data that are used to check a verifiable presentation requested from a holder;
+- *revocation authorities* can revoke credentials by signing a revocation message.
+
+Each contract instance MUST store public data of VCs of the same type.
+A *credential type* is a string that corresponds to the name of the VCs JSON schema, which credentials in the registry are based on.
+A `VCs JSON schema <https://w3c.github.io/vc-json-schema/>`_ is a `JSON schema <http://json-schema.org/>`_ describing the attributes of a VC.
+Attributes are sequentially numbered and have their numbers recorded in the additional ``index`` field in the schema.
+See VC schema examples and the corresponding VCs `here <https://github.com/Concordium/concordium-web3id/tree/main/examples/json-schemas>`_.
+
+.. TODO: refer to the Concordium VC Data Model documentation, once we have it.
+
+Specification
+=============
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in :rfc:`2119`.
+
+General types and serialization
+-------------------------------
+
+
+.. _CIS-4-Bool:
+
+``Bool``
+^^^^^^^^
+
+A boolean is serialized as a byte with value 0 for false and 1 for true::
+
+  Bool ::= (0: Byte) // False
+         | (1: Byte) // True
+
+
+.. _CIS-4-CredentialHolderId:
+
+``CredentialHolderId``
+^^^^^^^^^^^^^^^^^^^^^^
+
+A credential identifier is a holder's public key.
+It is expected that for each issued credential a unique public key is generated.
+The credential identifier also identifies the credential holder and can be used to prove ownership of the credential by the holder.
+
+It is serialized as :ref:`CIS-4-PublicKeyEd25519`.
+
+
+.. _CIS-4-MetadataUrl:
+
+``MetadataUrl``
+^^^^^^^^^^^^^^^
+
+A URL and optional checksum for metadata stored outside of this contract.
+
+It is serialized as: 2 bytes for the length of the metadata url (``n``) and then this many bytes for the url to the metadata (``url``) followed by an optional checksum.
```

Comment:
```suggestion
It is serialized as: 2 bytes for the length of the metadata URL (``n``) and then this many bytes for the URL to the metadata (``url``) followed by an optional checksum.
```

Reply:
Do you want to mention the endianness of `n`?

Reply:
Yes, this is a good point! I wonder why it's not mentioned in CIS-2, where I copied the definition from. And what is the endianness we actually use when serializing/deserializing the size?

Reply:
Clarified with @abizjak, it's little endian. I think I can write a general remark about how the size is serialized, because it's used in several places.